iroh-net/src/discovery.rs

```diff
@@ -95,6 +100,96 @@ pub trait Discovery: std::fmt::Debug + Send + Sync {
     ) -> Option<BoxStream<Result<DiscoveryItem>>> {
         None
     }
+
+    /// Subscribe to all addresses that get discovered.
```

Comment:
FWIW: at the end of this review I think I'm fine with keeping the subscribe method, but I do think it should only carry unsolicited `DiscoveryItem`s, so that most discovery mechanisms can just return `None`.

Reply:
Yes, this is a great question and not one I have a straight forward answer to.

I'm actually intrigued by your suggestion: only worry about adding events on a `subscribe` stream that were not explicitly asked for by calling `resolve`. I agree, this would make things much simpler for nodes that only discover "on command".

My question is: is this too weird? What I mean is, let's say you are writing a new discovery service, does this flexibility to report discovery in multiple ways make it confusing, or does this enable different "flavours" of discovery in a good way?

The more I think about it, the more I'm thinking it could be good if we give enough explanations and guidance, which I would try to do in the docs.

Reply:
The original design was that `resolve` would give you a stream of possible answers, once you established a connection you'd drop the stream to signal that you don't need any more answers.  I think part of the reasoning here was that maybe one discovery mechanism would yield a working result and then the others can stop expensive work.  

The new requirement is to support discovery mechanisms that just unilaterally push node information to the endpoint.  I think I'm in favour of clearly making them for different purposes.  If we clearly document that nodes discovered for `resolve` SHOULD not be pushed to the subscribers then this should not be that weird.  I think it will make the reason to have both these mechanisms much clearer.

So yes, even 24h later I'm still in favour of this.